lib/webauth.js

```diff
@@ -125,10 +131,60 @@ StatusResponse.prototype = {
   keyField: "authenticated"
 }
 
+const SESSION_ACTION_TYPE_AUTHENTICATE = 1;
+const SESSION_ACTION_TYPE_REFRESH = 2;
+
 /*
  * Assumes req.session is there and behaves as it should
  */
 module.exports = function(authManager) {
+  const _authenticateOrRefresh = Promise.coroutine(function*(req, res, type) {
+    let functionName;
+    if (type == SESSION_ACTION_TYPE_AUTHENTICATE) {
+      functionName = 'authenticate';
+    } else if (type == SESSION_ACTION_TYPE_REFRESH) {
+      functionName = 'refreshStatus';
+    } else {
+      res.status(500).json({error: "Invalid session action type attempted"});
+      return;
+    }
+    
+    try {
+      const result = new LoginResult();
+      const handlers = getRelevantHandlers(authManager, req.body);
+      const authServiceHandleMaps = 
+            req[`${UNP.APP_NAME}Data`].webApp.authServiceHandleMaps;
+      for (const handler of handlers) {
+        const pluginID = handler.pluginID;
+        const authPluginSession = getAuthPluginSession(req, pluginID, {});
+        req[`${UNP.APP_NAME}Data`].plugin.services = 
+          authServiceHandleMaps[pluginID];
+        // FIXME This is a bug: in webauth.js we shouldn't make assumptions
+        // about the contents of the session. We only can look at the return 
+        // values of the method.          
+        const wasAuthenticate = authPluginSession.authenticated;
```

Comment:
It should be `wasAuthenticated`

Reply:
This area seems odd but functional. The result of getStatus returns authenticated, while other actions return success, while in this block essentially success = authenticated. Do you remember why you wanted this distinction? I fixed the conditional slightly below since it seemed to have the same issue as your FIXME comment.

Reply:
My idea is that the contract for auth plugins is the methods and their params/return values. The contents of the session, on the other hand, is private to the plugin, and is totally opaque to the framework.

Here, the framework made assumptions about the contents of the session, based on what two existing plugins do. If this had not been fixed, all other future plugins would've been forced to follow the same convention and also have a boolean field called "authenticated" on their session, and to maintain it properly.

And BTW, even in this case, who would've checked the session expiration?